solutions/CSES Planets Cycles.mdx

```diff
@@ -0,0 +1,209 @@
+---
+id: cses-1751
+source: CSES 
+title: Planets Cycles
+author: Maggie Liu (C++), Qi Wang (Java)
+---
+
+Define the number of teleportations starting from a planet as the $\texttt{pathlength}$ of that planet. For each planet that hasn't been visited, we want to find its $\texttt{pathlength}$. Call the planet we are performing $\texttt{dfs}$ from the $\textit{start}$. As we perform $\texttt{dfs}$ from the $\textit{start}$, keep track of the planets seen, in order, in the $\texttt{path}$ queue and keep track of the $\texttt{steps}$, the length of the path so far. When we reach a planet that has already been visited (call this planet the $\textit{repeat}$), add the $\texttt{pathlength}$ of the $\textit{repeat}$ to the current $\texttt{step}$ count because we would continue to visit all of the planets that $\textit{repeat}$ would go on to visit. 
+
+Once we have $\texttt{path}$ and $\texttt{steps}$ (the $\texttt{pathlength}$ of the $\textit{start}$) we can calculate the $\texttt{pathlength}$ of all the planets in this $\texttt{path}$. We know the $\textit{repeat}$ will always be the planet at the end of the $\texttt{path}$, but it may appear elsewhere as well. We can break this down into two cases:
+1. The $\textit{repeat}$ was visited twice in the current $\texttt{path}$. Let the planets in the $\texttt{path}$ between the two occurrences of the $\textit{repeat}$ be the $\textit{cycle}$.
+2. The $\textit{repeat}$ was visited previously and only appears at the end of the current $\texttt{path}$. All of the planets in the $\texttt{path}$ are not part of a $\textit{cycle}$.
+
+For planets inside a $\textit{cycle}$, the repeating planet when starting from that planet is itself. For all the planets outside the $\textit{cycle}$, the planet that repeats when starting from each planet will still be the $\textit{repeat}$.
+
+Since the planets outside the $\textit{cycle}$ all have paths ending at the $\textit{repeat}$, each one's $\texttt{pathlength}$ is $1$ less than the previous'. So, as we iterate through the planets outside the $\textit{cycle}$, the $\texttt{pathlength}$ will decrease by $1$ each time, starting from $\texttt{steps}$. Once we get to the $\textit{cycle}$, the $\texttt{pathlength}$ of the planets will all be equal to the $\texttt{pathlength}$ of the $\textit{repeat}$, which is the length of the $\textit{cycle}$.
+
+<LanguageSection>
+
+<CPPSection>
+
+## C++ Implementation
+
+```cpp
+#include <iostream>
+#include <queue>
+using namespace std;
+void dfs(int planet);
+
+bool visited[200000] = {false};
```

Comment:
`int pathlength[200000] = {0};`

global arrays are already initialized to zero. Also, this may give the impression that this sets the entire array to `0`. The following would set the first element to 1 and the remaining elements to zero:

`int pathlength[200000] = {1};`

Would prefer

`int pathlength[200000]{};`

as mentioned here: https://www.learncpp.com/cpp-tutorial/arrays-part-ii/